### PR TITLE
Recovery: Expanding the daemon suspicious_replica_recoverer #4565

### PR DESCRIPTION
**Pull request by ChristophAmes**
The idea is to expand the daemon "suspicious_replica_recoverer", so that it goes into more detail when dealing with suspicious files.

The first main change is that it can now deal with both suspicious replicas that have additional copies on other RSEs and replicas that are the last remaining copy, the latter being previously ignored. The last remaining copies are dealt with differently based on the file type / metadata of the replica:
Log files and user files are declared bad without further investigation. Replicas with the metadata " datatype='RAW' " are not declared bad, but are instead ignored (meaning that they will have to be dealt with manually).
The rest of the file types are currently being ignored, to make sure no mistakes are made. This can be easily changed by adding new entries to suspicious_replica_recoverer.json.

To be able to make these changes, the function "get_suspicious_replicas" from lib/rucio/core/replica.py had to be changed. The parameter "available_elsewhere" now has three options:
0: All suspicious replicas
1: Suspicious replicas that have at least one copy on another RSE
2: Suspicious replicas that are the last remaining copy

The second main change is that RSEs with more than a certain number of suspicious replicas (currently set to 5, but requires testing when the daemon is run in passive mode) are considered "problematic". This means that it is assumed that the high number of suspicious replicas is due to the RSE having a problem, as opposed to the replicas themselves having problems. In such a case, the suspicious replicas are declared temporarily unavailable.

### CODE REVIEW

File: bin/rucio-replica-recoverer [hunk 1]
```diff
@@ -235,13 +238,11 @@ Note that attempting the use the ``--vos`` argument when in single-VO mode will
...
+    parser.add_argument("--nattempts", action="store", default=10, help='Minimum count of suspicious file replica appearance in bad_replicas table. Default value is 10.')
     parser.add_argument("--younger-than", action="store", default=3, help='Consider all file replicas logged in bad_replicas table since speicified number of younger-than. Default value is 3.')
-    parser.add_argument("--rse-expression", action="store", default='MOCK', help='The RSE on which the recovery should be happening.')
     parser.add_argument('--vos', nargs='+', type=str, help='Optional list of VOs to consider. Only used in multi-VO mode.')
     parser.add_argument("--run-once", action="store_true", default=False, help='One iteration only.')
-    parser.add_argument("--max-replicas-per-rse", action="store", default=100, help='The maximum number of suspicious replicas found on an RSE which may be declared bad. If a higher count is found, no action is be taken.')
-    parser.add_argument('--sleep-time', action="store", default=60, type=int, help='Concurrency control: thread sleep time after each chunk of work')
```
Comment (cserf) [inline bin/rucio-replica-recoverer:244]: Why did you remove `--sleep-time` ?  Can you put it back please

File: lib/rucio/core/replica.py [hunk 2]
```diff
@@ -3168,9 +3171,19 @@ def get_suspicious_files(rse_expression, filter_=None, **kwargs):
     nattempts = kwargs.get("nattempts", 0)
     session = kwargs.get("session", None)
     exclude_states = kwargs.get("exclude_states", ['B', 'R', 'D'])
-    available_elsewhere = kwargs.get("available_elsewhere", False)
+    available_elsewhere = kwargs.get("available_elsewhere", 0)
     is_suspicious = kwargs.get("is_suspicious", False)
 
+    if available_elsewhere not in [0, 1, 2]:
+        print("""ERROR, available_elsewhere must be set to one of the following:
```
Comment (cserf) [inline lib/rucio/core/replica.py]: Use a logger instead (see https://github.com/rucio/rucio/blob/edbad7d45262eda611f19a69e1ab1cb726098764/lib/rucio/core/replica.py#L945)

File: lib/rucio/core/replica.py [hunk 3]
```diff
@@ -3223,7 +3246,7 @@ def get_suspicious_files(rse_expression, filter_=None, **kwargs):
     # finally, the results are grouped by RSE, scope, name and required to have
     # at least 'nattempts' occurrences in the result of the query per replica
     query_result = query.group_by(models.RSEFileAssociation.rse_id, bad_replicas_alias.scope, bad_replicas_alias.name).having(func.count() > nattempts).all()
-    # print(query)
```
Comment (cserf) [inline lib/rucio/core/replica.py:3264]: Remove

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 4]
```diff
@@ -40,30 +42,34 @@
...
 from rucio.common.types import InternalAccount
-from rucio.common.utils import daemon_sleep
 from rucio.core.heartbeat import live, die, sanity_check
 from rucio.core.monitor import record_counter
-from rucio.core.replica import list_replicas, declare_bad_file_replicas, get_suspicious_files
+from rucio.core.did import get_metadata
+from rucio.core.replica import list_replicas, get_suspicious_files, add_bad_pfns, declare_bad_file_replicas
 from rucio.core.rse_expression_parser import parse_expression
+
+# from rucio.core.rse import list_rses
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Remove

File: lib/rucio/tests/test_replica_recoverer.py [hunk 5]
```diff
@@ -169,7 +173,8 @@ def test_replica_recoverer(self):
 
         # Run replica recoverer once
         try:
-            run(once=True, younger_than=1, nattempts=2, rse_expression='MOCK_SUSPICIOUS')
+            run(once=True, younger_than=1, nattempts=2, limit_suspicious_files_on_rse=5)
+            print("Run")
```
Comment (cserf) [inline lib/rucio/tests/test_replica_recoverer.py]: Remove the print statement

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 6]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
 
             start = time.time()
 
-            logging.info('replica_recoverer[%i/%i]: ready to query replicas at RSE %s,'
-                         + ' reported suspicious in the last %i days at least %i times which are available on other RSEs.',  # NOQA: W503
-                         worker_number, total_workers, rse_expression, younger_than, nattempts)
+            logging.info('replica_recoverer[%i/%i]: Ready to query replicas which were'
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Use a logger (see in https://github.com/rucio/rucio/blob/edbad7d45262eda611f19a69e1ab1cb726098764/lib/rucio/daemons/reaper/reaper.py for instance)

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 7]
```diff
@@ -103,23 +109,6 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
             vos = [v['vo'] for v in list_vos()]
         logging.info('replica_recoverer: This instance will work on VO%s: %s' % ('s' if len(vos) > 1 else '', ', '.join([v for v in vos])))
 
-    # Don't require a result from the expression at each VO, only raise if we can't get a result from any of them
-    rses = []
-    exceptions_raised = 0
-    for vo in vos:
-        try:
-            parsed_rses = parse_expression(expression=rse_expression, filter_={'vo': vo})
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py:113]: If `rse_expression` is not used anymore, maybe it can be removed from the method signature

File: lib/rucio/tests/test_replica_recoverer.py [hunk 8]
```diff
@@ -169,7 +173,8 @@ def test_replica_recoverer(self):
 
         # Run replica recoverer once
         try:
-            run(once=True, younger_than=1, nattempts=2, rse_expression='MOCK_SUSPICIOUS')
+            run(once=True, younger_than=1, nattempts=2, limit_suspicious_files_on_rse=5)
```
Comment (cserf) [inline lib/rucio/tests/test_replica_recoverer.py:210]: I would add 1 or 2 other tests :
- 1 with RAW data and check that the daemon always ignore it
- 1 with datatype not specified in the json. For now it should ignore it
- 1 with a datatype specified and action `declare_bad`.

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 9]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
-            logging.info('replica_recoverer[%i/%i]: ready to query replicas at RSE %s,'
-                         + ' reported suspicious in the last %i days at least %i times which are available on other RSEs.',  # NOQA: W503
-                         worker_number, total_workers, rse_expression, younger_than, nattempts)
+            logging.info('replica_recoverer[%i/%i]: Ready to query replicas which were'
+                         + ' reported as suspicious in the last %i days at least %i times.',
+                         worker_number, total_workers, younger_than, nattempts)
+
+            getfileskwargs_avail_elsewhere = {'younger_than': younger_than,
+                                              'nattempts': nattempts,
+                                              'exclude_states': ['B', 'R', 'D', 'L', 'T'],
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: I wonder if we should consider `T` i.e. temporary available in the list of states to exclude. To be discussed with Rod

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 10]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
-                                            worker_number, total_workers, max_replicas_per_rse, str(len(surls_to_recover[vo][rse_id])), rse)
-                        else:
-                            declare_bad_file_replicas(pfns=surls_to_recover[vo][rse_id], reason='Suspicious. Automatic recovery.', issuer=InternalAccount('root', vo=vo), status=BadFilesStatus.BAD, session=None)
-                            logging.info('replica_recoverer[%i/%i]: finished declaring bad replicas on %s.', worker_number, total_workers, rse)
+                logging.info('replica_recoverer[%i/%i]: Start replica recovery for VO: %s', worker_number, total_workers, vo)
+                recoverable_replicas = {}
+                if vo not in recoverable_replicas:
+                    recoverable_replicas[vo] = {}
+
+                rse_list = sorted([rse for rse in parse_expression('enable_suspicious_file_recovery=true')], key=lambda k: k['rse'])
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Need to add a VO filter since there is a loop on VO

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 11]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
+
+                logging.debug("List of RSEs with enable_suspicious_file_recovery = True:")
+                for i in rse_list:
+                    logging.debug(i)
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Put the prepend string. Done automatically when you use the logger
Comment valid for all the other calls to logging

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 12]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
+                    rse_expr = rse['rse']
+                    cnt_surl_not_found = 0
+                    if rse_expr not in recoverable_replicas[vo]:
+                        recoverable_replicas[vo][rse_expr] = {}
+                    # Get a dictionary of the suspicious replicas on the RSE that have available copies on other RSEs
+                    suspicious_replicas_avail_elsewhere = get_suspicious_files(rse_expr, filter={'vo': vo}, **getfileskwargs_avail_elsewhere)
+                    # Get the suspicious replicas that are the last remaining copies
+                    suspicious_replicas_last_copy = get_suspicious_files(rse_expr, filter={'vo': vo}, **getfileskwargs_last_copy)
+
+                    logging.debug('\n')
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Logs are sent to elasticSearch. Try to avoid putting empty lines

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 13]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
+                                    logging.warning('replica_recoverer[%i/%i]: Skipping suspicious replica %s on %s, no surls were found.', worker_number, total_workers, rep_name, rse_expr)
+
+                    if suspicious_replicas_last_copy:
+                        for replica in suspicious_replicas_last_copy:
+                            if vo == replica['scope'].vo:
+                                scope = replica['scope']
+                                rep_name = replica['name']
+                                rse_id = replica['rse_id']
+                                surl_not_found = True
+                                for rep in list_replicas([{'scope': scope, 'name': rep_name}]):
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py:231]: I'm not sure I understand this check. Since it is the last replica, you should never find a valid replica.

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 14]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
+                logging.info('\n\n\n\n\n\n\n\n')
+                logging.info('replica_recoverer[%i/%i]: All RSEs have been checked for suspicious replicas. Total time: %.2f seconds.', worker_number, total_workers, time.time() - start)
+                logging.info('replica_recoverer[%i/%i]: Begin check for problematic RSEs.\n\n\n', worker_number, total_workers)
+                time_start_check_probl = time.time()
+
+                # If an RSE has more than *limit_suspicious_files_on_rse* suspicious files, then there might be a problem with the RSE.
+                # The suspicious files are marked as temporarily unavailable.
+                list_problematic_rses = []
+                for rse_key in list(recoverable_replicas[vo].keys()):
+                    if len(recoverable_replicas[vo][rse_key].values()) > limit_suspicious_files_on_rse:
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py:256]: I wonder if we shouldn't put an upper limit too. If there are millions of files, we shouldn't probably declare them temporary unavailable, but just set the site to unavailable and notify the site.
But that's second order correction that can be added later

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 15]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
+                            logging.info('replica_recoverer[%i/%i]: Scope: %s    Name: %s', worker_number, total_workers, rse_values['scope'], rse_values['name'])
+                        # Remove the RSE from the dictionary as it has been dealt with.
+                        del recoverable_replicas[vo][rse_key]
+
+                logging.info("\n\n")
+                logging.info("replica_recoverer[%i/%i]: Following RSEs were deemed problematic (total: %i)", worker_number, total_workers, len(list_problematic_rses))
+                for rse in list_problematic_rses:
+                    logging.info("replica_recoverer[%i/%i]: %s", worker_number, total_workers, rse)
+
+                json_file = open("/opt/rucio/etc/suspicious_replica_recoverer.json")
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Just do it at the beginning of the loop on VOs.

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 16]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
+                json_data = json.load(json_file)
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Add a try except (if the json is badly formated it will crash). Also a validation of the json (check if mandatory fields are available can be useful)

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 17]
```diff
@@ -145,75 +134,190 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
+                    # Get the rse_id by going to one of the suspicious replicas from that RSE and reading it from there
+                    rse_id = list(recoverable_replicas[vo][rse_key].values())[0]['rse_id']
+                    for replica_key in list(recoverable_replicas[vo][rse_key].keys()):
+                        if recoverable_replicas[vo][rse_key][replica_key]['available_elsewhere'] is True:
+                            # Replicas with other copies on at least one other RSE can safely be labeled as bad
+                            files_to_be_declared_bad.append(recoverable_replicas[vo][rse_key][replica_key]['surl'])
+                            # Remove replica from dictionary
+                            del recoverable_replicas[vo][rse_key][replica_key]
+                        elif recoverable_replicas[vo][rse_key][replica_key]['available_elsewhere'] is False:
+                            if (recoverable_replicas[vo][rse_key][replica_key]['name'].startswith("log.")) or (recoverable_replicas[vo][rse_key][replica_key]['name'].startswith("user")):
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py:293]: ATLAS specific policy. Will have to be modified if used by another VO. OK to keep like this for now

File: bin/rucio-replica-recoverer [hunk 18]
```diff
@@ -235,13 +238,12 @@ Note that attempting the use the ``--vos`` argument when in single-VO mode will
...
+    parser.add_argument("--limit-suspicious-files-on-rse", action="store", default=5, help='Maximum number of suspicious replicas on an RSE before that RSE is considered problematic and the suspicious replicas on that RSE are declared "TEMPORARY_UNAVAILABLE". Default value is 5.')
+    parser.add_argument('--sleep-time', action="store", default=300, type=int, help='Concurrency control: thread sleep time after each chunk of work')
```
Comment (mlassnig) [inline bin/rucio-replica-recoverer:247]: 60 to 300?

File: bin/rucio-replica-recoverer [hunk 19]
```diff
@@ -208,9 +211,9 @@ $$ from datetime import datetime, timedelta
 $$ from_date = datetime.now() - timedelta(days=3)
 $$ from rucio.core.replica import list_bad_replicas_status
 
-$$ get_suspicious_files('MOCK_SUSPICIOUS',younger_than=from_date, nattempts=10, is_suspicious=True, available_elsewhere=True)
+$$ get_suspicious_files('MOCK_SUSPICIOUS',younger_than=from_date, nattempts=10, is_suspicious=True, available_elsewhere=1)
```
Comment (mlassnig) [inline bin/rucio-replica-recoverer]: Can we change this? It's not really clear what the numbers mean

`
0: All suspicious replicas
1: Suspicious replicas that have at least one copy on another RSE
2: Suspicious replicas that are the last remaining copy
`

maybe better to add something to lib/rucio/common/constants?

SUSPICIOUS.ALL
SUSPICIOUS.EXISTING_REPLICAS
SUSPICIOUS.LAST_REPLICA

File: bin/rucio-replica-recoverer (hunk 18 above)
Comment (mlassnig) [inline bin/rucio-replica-recoverer:247]: 60 to 300?

Comment (ChristophAmes) [reply bin/rucio-replica-recoverer:247]: The daemon generally takes up to five minutes to run, since it now runs over all RSEs at once, instead of going over them individually. It doesn't make sense to have sleep_time set to 60, as it will always take longer. I did however just notice that I forgot to include daemon_sleep, so I just created a new commit and pushed.

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 20]
```diff
@@ -41,30 +42,34 @@
...
 from rucio.core.vo import list_vos
-from rucio.db.sqla.constants import BadFilesStatus
 from rucio.db.sqla.util import get_db_time
 
 GRACEFUL_STOP = threading.Event()
 
 
-def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rse_expression='MOCK', vos=None,
-                                    max_replicas_per_rse=100, sleep_time=60):
+# logging.basicConfig(filename='example.log', filemode='w', level=logging.DEBUG)
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: Remove

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 21]
```diff
@@ -136,112 +128,237 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
+
             start = time.time()
 
-            logging.info('replica_recoverer[%i/%i]: ready to query replicas at RSE %s,'
-                         + ' reported suspicious in the last %i days at least %i times which are available on other RSEs.',  # NOQA: W503
-                         worker_number, total_workers, rse_expression, younger_than, nattempts)
+            try:
+                json_file = open("/opt/rucio/etc/suspicious_replica_recoverer.json")
+            except:
+                logger(logging.DEBUG, "An error occured whilst trying to open the json file.")
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: WARNING or ERROR, but not DEBUG

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 22]
```diff
@@ -136,112 +128,237 @@ def declare_suspicious_replicas_bad(once=False, younger_than=3, nattempts=10, rs
...
+                break
+
+            json_data = json.load(json_file)
```
Comment (cserf) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: You should put a try except and catch `ValueError: No JSON object could be decoded`

**Review by cserf — DISMISSED** [review]
Just a few cosmetic changes but I think it's good to go.

**Review by cserf — DISMISSED** [review]

File: bin/rucio-replica-recoverer (hunk 19 above)
Comment (bari12) [reply bin/rucio-replica-recoverer]: This should be added with the constants then to the doc/code as well

File: lib/rucio/core/replica.py [hunk 23]
```diff
@@ -3136,7 +3138,7 @@ def get_replicas_state(scope=None, name=None, session=None):
 
 
 @read_session
-def get_suspicious_files(rse_expression, filter_=None, **kwargs):
+def get_suspicious_files(rse_expression, filter_=None, logger=logging.log, **kwargs):
```
Comment (bari12) [inline lib/rucio/core/replica.py]: Any reason why `kwargs` is used here? We should really try to avoid this for core functions.

**Review by bari12 — COMMENTED** [review]
Hi @ChristophAmes 
Very sorry for the late comments here. @cserf just asked me about this PR and I did not realise that it is still unmerged. Just two comments.

**Comment by bari12** [comment]
Please also resolve the comments you already addressed.

File: lib/rucio/core/replica.py (hunk 23 above)
Comment (ChristophAmes) [reply lib/rucio/core/replica.py]: The documentation needs a few changes, since the daemon has a different structure. I'll have to do that in the future.

Comment (bari12) [reply lib/rucio/core/replica.py]: Ok, for me it's fine to do this later, please create a ticket for it though and create a PR later on 😄

**Review by cserf — APPROVED** [review]

**Comment by bari12** [comment]
Hi,
I am merging this like this to have it in the release, but there is a few things which should be addressed:
- At the core level, do not deal with the constant values anymore, here you should assume (and ensure) that it is the actual constants handed in to the core. (I know there is a few other parts in the code where this is not the case, but these are also not clean, and we should get away from that)
- In the daemon itself there is quite a few cases where you loop over replicas and then do individual `list_replicas` calls. That's not very efficient and these should be changed to bulk calls where possible.

Best,
Martin
